File: python/tournament_winner/tournament_winner.py
# O(n) time | O(k) space;
# where n is the number of competitions and k is the number of teams
def tournamentWinner(competitions, results):
    # Initialize the current best team and the scores dictionary
    currentBestTea = ""
    scores = {currentBestTeam: 0}

    # Iterate through each competition and its result
    for idx, competition in enumerate(competitions):
        # Extract the home team and away team from the competition
        homeTeam, awayTeam = competition
        # Get the result of the current competition
        result = results[idx]

        # Determine the winning team based on the result
        winningTeam = homeTeam if result == 1 else awayTeam

        # Initialize the winning team's score if not already present
        if winningTeam not in scores:
            scores[winningTeam] = 0

        # Add 3 points to the winning team's score
        scores[winningTeam] += 3

        # Update the current best team if the winning team's score is higher
        if scores[winningTeam] > scores[currentBestTeam]:
            currentBestTeam = winningTeam

    # Return the team with the highest score
    return currentBestTeam

File: java/tournament_winner/TournamentWinner.java
package main.java.com.example.tournamentwinner;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;

public class TournamentWinner {


	// O(n) time | O(k) space;
	// where n is the length of arrayOne and m is the length of arrayTwo
	public static String tournamentWinner(ArrayList<ArrayList<String>> competitions, ArrayList<Integer> results) {
        // Initialize the current best team with an empty string and create a map to keep track of team scores
        String currentBestTeam = "";
        Map<String, Integer> scores = new HashMap<>();
        // Set the initial score of the current best team (empty string) to 0
        scores.put(currentBestTeam, 0);
        
        // Iterate through each competition
        for (int i = 0; i < competitions.size(); i++) {
            // Retrieve the current competition's teams and result
            ArrayList<String> competition = competitions.get(i);
            String homeTeam = competition.get(0);
            String awayTeam = competition.get(1);
            int result = results.get(i);
            
            // Determine the winning team based on the result
            String winningTeam = result == 1 ? homeTeam : awayTeam;
    
            // Update the score for the winning team by adding 3 points
            scores.put(winningTeam, scores.getOrDefault(winningTeam, 0) + 3);

            // Check if the current winning team's score is higher than the current best team's score
            if (scores.get(winningTeam) > scores.get(currentBestTeam)) {
                // Update the current best team to the new winning team
                currentBestTeam = winningTeam;
            }
        }

        // Return the name of the team with the highest score
        return currentBestTeam;
    }
}